app/controllers/application_controller.rb

```diff
@@ -5,6 +5,103 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "password_security"
+  end
+
+
+  get '/' do
+    erb :'index'
+  end
+
+
+  get '/signup' do
+    erb :'users/create_user'
+  end
+
+  post '/signup' do
+    if params[:username] == "" || params[:password] == "" || params[:email] == ""
+      redirect '/failure'
+    else
+      @user =User.create(username: params[:username], email: params[:email], password: params[:password])
+      session[:user_id] = @user.id
+      redirect '/tweets'
+    end
+  end
+
+  get '/login' do
+    erb :'users/login'
+  end
+
+  post "/login" do
+    @user = User.find_by(username: params[:username])
+    if @user && @user.authenticate(params[:password])
+      session[:user_id] = @user.id
+      redirect to "/tweets"
+    else
+      redirect to "/failure"
+    end
+  end
+
+    get '/tweets' do
+      erb :'tweets/tweets'
+    end
+
+    get "/logout" do
+    session.clear
+    redirect "/"
+  end
+
+  get '/tweets/new' do
+    erb :'tweets/create_tweet'
+  end
+
+  post '/tweets' do
+    if !logged_in?
+      redirect to "/failure"
+    else
+      tweet = current_user.tweets.build(params)
+      tweet.save
+    end
+    redirect to '/tweets'
+  end
+
+  get '/tweets/:id' do
+      @tweet = Tweet.find(params[:id])
+      erb :'tweets/show'
+    end
+
+  get '/tweets/:id/edit' do
+    @tweet = Tweet.find(params[:id])
+    erb :'tweets/edit'
+  end
+
+  post '/tweets/:id' do
+    tweet = Tweet.find(params[:id])
+     tweet.update(content: params[:content])
+     redirect to '/tweets'
+  end
+
+  post '/tweets/:id/delete' do
+    tweet = Tweet.find(params[:id])
+     tweet.destroy
+     redirect to '/tweets'
+  end
+
+
+
+
+
+
+
+   helpers do
+    def logged_in?
+      !!session[:user_id]
```

Comment:
It's better to rely on the successful call to `current_user` than only introspecting on the `session[:id]` value. In theory, a user might have an invalid `id` in their session which would make `logged_in?` return true even though that user would not be findable later. Also, because `current_user` is memoized, it makes sense to call it as early in the request cycle as possible, which generally is the first call `logged_in?`.

```ruby
def logged_in?
  !!current_user
end
```